File: experiments/whenwords-ref-python/whenwords.py
"""whenwords — 5 pure functions for human-friendly date/time formatting."""

import math
import re
from datetime import datetime, timezone


def _js_round(x: float) -> int:
    """Round half-up (JavaScript-style), not half-to-even (Python default)."""
    return math.floor(x + 0.5)


def time_ago(timestamp: int, reference: int) -> str:
    """Return a human-friendly relative time string like '5 minutes ago' or 'in 3 hours'."""
    seconds = abs(reference - timestamp)
    is_future = timestamp > reference

    # (max_seconds, singular, plural, divisor)
    thresholds: list[tuple[float, str, str, int]] = [
        (44, "just now", "just now", 1),
        (89, "1 minute", "1 minute", 1),
        (44 * 60, "minute", "minutes", 60),
        (89 * 60, "1 hour", "1 hour", 1),
        (21 * 3600, "hour", "hours", 3600),
        (35 * 3600, "1 day", "1 day", 1),
        (25 * 86400, "day", "days", 86400),
        (45 * 86400, "1 month", "1 month", 1),
        (319 * 86400, "month", "months", 30 * 86400),
        (547 * 86400, "1 year", "1 year", 1),
        (math.inf, "year", "years", 365 * 86400),
    ]

    max_s, singular, plural, divisor = next(
        (m, s, p, d) for m, s, p, d in thresholds if seconds <= m
    )

    if singular == "just now":
        return "just now"

    if divisor == 1:
        return f"in {singular}" if is_future else f"{singular} ago"

    n = _js_round(seconds / divisor)
    label = singular if n == 1 else plural
    text = f"{n} {label}"
    return f"in {text}" if is_future else f"{text} ago"


def duration(seconds: int, *, compact: bool = False, max_units: int = 2) -> str:
    """Format a duration in seconds as a human-readable string."""
    if seconds < 0:
        raise ValueError("Seconds must not be negative")

    units: list[tuple[str, str, int]] = [
        ("year", "y", 365 * 86400),
        ("month", "mo", 30 * 86400),
        ("day", "d", 86400),
        ("hour", "h", 3600),
        ("minute", "m", 60),
        ("second", "s", 1),
    ]

    if seconds == 0:
        return "0s" if compact else "0 seconds"

    parts: list[str] = []
    remaining = seconds

    for name, abbr, size in units:
        if remaining >= size:
            is_last_unit = len(parts) + 1 >= max_units
            count = _js_round(remaining / size) if is_last_unit else remaining // size
            remaining = 0 if is_last_unit else remaining % size

            if count > 0:
                if compact:
                    parts.append(f"{count}{abbr}")
                else:
                    label = name if count == 1 else f"{name}s"
                    parts.append(f"{count} {label}")

        if len(parts) >= max_units:
            break

    return " ".join(parts) if compact else ", ".join(parts)


def parse_duration(input_str: str) -> int:
    """Parse a human-readable duration string into total seconds.

    Supports compact ('2h30m'), verbose ('2 hours 30 minutes'),
    colon ('2:30'), decimal ('2.5 hours'), and mixed formats.
    """
    text = input_str.strip()
    if not text:
        raise ValueError("Empty input")

    # Check for negatives
    if text.startswith("-"):
        raise ValueError("Negative durations are not allowed")

    # Try colon format first: H:MM or H:MM:SS
    colon_match = re.fullmatch(r"(\d+):(\d{1,2})(?::(\d{1,2}))?", text)
    if colon_match:
        hours = int(colon_match.group(1))
        minutes = int(colon_match.group(2))
        secs = int(colon_match.group(3)) if colon_match.group(3) else 0
        return hours * 3600 + minutes * 60 + secs

    unit_map: dict[str, int] = {
        "w": 604800, "week": 604800, "weeks": 604800,
        "d": 86400, "day": 86400, "days": 86400,
        "h": 3600, "hr": 3600, "hrs": 3600, "hour": 3600, "hours": 3600,
        "m": 60, "min": 60, "mins": 60, "minute": 60, "minutes": 60,
        "mo": 2592000, "month": 2592000, "months": 2592000,
        "s": 1, "sec": 1, "secs": 1, "second": 1, "seconds": 1,
    }

    # Match all number+unit pairs
    # Use lookahead instead of \b so "2h30m" works (no word boundary between "h" and "3")
    pattern = re.compile(
        r"(\d+(?:\.\d+)?)\s*(weeks?|w|months?|mo|days?|d|hrs?|hours?|h|mins?|minutes?|m|secs?|seconds?|s)(?=\s|,|\d|$)",
        re.IGNORECASE,
    )

    # Strip filler words for matching
    cleaned = re.sub(r"\b(and|,)\b", " ", text, flags=re.IGNORECASE)
    cleaned = re.sub(r",", " ", cleaned)

    matches = pattern.findall(cleaned)
    if not matches:
        raise ValueError(f"Unrecognized duration format: {input_str!r}")

    # Verify that we consumed the entire meaningful input (no bare numbers, no unknown units)
    consumed = pattern.sub("", cleaned).strip()
    # Remove commas, "and", whitespace from remainder
    remainder = re.sub(r"[,\s]+", "", consumed)
    remainder = re.sub(r"(?i)\band\b", "", remainder)
    if remainder:
        raise ValueError(f"Unrecognized duration format: {input_str!r}")

    total = 0.0
    for value_str, unit in matches:
        value = float(value_str)
        unit_lower = unit.lower()
        if unit_lower not in unit_map:
            raise ValueError(f"Unknown unit: {unit!r}")
        total += value * unit_map[unit_lower]

    return _js_round(total)


def human_date(timestamp: int, reference: int) -> str:
    """Return a human-friendly date label relative to a reference date.

    Returns 'Today', 'Yesterday', 'Tomorrow', 'Last Wednesday',
    'This Wednesday', 'March 5', or 'January 1, 2023'.
    """
    ts_dt = datetime.fromtimestamp(timestamp, tz=timezone.utc)
    ref_dt = datetime.fromtimestamp(reference, tz=timezone.utc)

    ts_date = ts_dt.date()
    ref_date = ref_dt.date()

    diff_days = (ts_date - ref_date).days  # positive = future

    if diff_days == 0:
        return "Today"
    if diff_days == -1:
        return "Yesterday"
    if diff_days == 1:
        return "Tomorrow"

    day_names = [
        "Monday", "Tuesday", "Wednesday", "Thursday",
        "Friday", "Saturday", "Sunday",
    ]

    # "Last <day>" for 2–6 days in the past
    if -6 <= diff_days <= -2:
        return f"Last {day_names[ts_date.weekday()]}"

    # "This <day>" for 2–6 days in the future
    if 2 <= diff_days <= 6:
        return f"This {day_names[ts_date.weekday()]}"

    # Different year
    month_names = [
        "January", "February", "March", "April", "May", "June",
        "July", "August", "September", "October", "November", "December",
    ]

    month_name = month_names[ts_date.month - 1]

    if ts_date.year != ref_date.year:
        return f"{month_name} {ts_date.day}, {ts_date.year}"

    # Same year
    return f"{month_name} {ts_date.day}"


def date_range(start: int, end: int) -> str:
    """Format a date range as a human-readable string.

    Uses en-dash for ranges. Auto-corrects swapped inputs.
    """
    if start > end:
        start, end = end, start

    start_dt = datetime.fromtimestamp(start, tz=timezone.utc)
    end_dt = datetime.fromtimestamp(end, tz=timezone.utc)

    start_date = start_dt.date()
    end_date = end_dt.date()

    month_names = [
        "January", "February", "March", "April", "May", "June",
        "July", "August", "September", "October", "November", "December",
    ]

    s_month = month_names[start_date.month - 1]
    e_month = month_names[end_date.month - 1]

    en_dash = "\u2013"

    # Same day
    if start_date == end_date:
        return f"{s_month} {start_date.day}, {start_date.year}"

    # Same month and year
    if start_date.year == end_date.year and start_date.month == end_date.month:
        return f"{s_month} {start_date.day}{en_dash}{end_date.day}, {start_date.year}"

    # Same year, different months
    if start_date.year == end_date.year:
        return (
            f"{s_month} {start_date.day} {en_dash} "
            f"{e_month} {end_date.day}, {start_date.year}"
        )

    # Different years
    return (
        f"{s_month} {start_date.day}, {start_date.year} {en_dash} "
        f"{e_month} {end_date.day}, {end_date.year}"
    )

File: research/experiments/sweep/runs/whenwords-full-csharp-sonnet45/WhenWords/WhenWords.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text.RegularExpressions;

namespace WhenWords
{
    public static class WhenWords
    {
        // ============================================================================
        // time-ago: Converts a Unix timestamp to a relative time string
        // ============================================================================

        public static string TimeAgo(long timestamp, long reference)
        {
            long seconds = Math.Abs(reference - timestamp);
            bool isFuture = timestamp > reference;

            // Threshold table: [maxSeconds, singularLabel, pluralLabel, divisor]
            var thresholds = new (long max, string singular, string plural, long divisor)[]
            {
                (44, "just now", "just now", 1),
                (89, "1 minute ago", "in 1 minute", 1),
                (2640, "minutes ago", "in minutes", 60),
                (5340, "1 hour ago", "in 1 hour", 1),
                (75600, "hours ago", "in hours", 3600),
                (126000, "1 day ago", "in 1 day", 1),
                (2160000, "days ago", "in days", 86400),
                (3888000, "1 month ago", "in 1 month", 1),
                (27561600, "months ago", "in months", 2592000),
                (47260800, "1 year ago", "in 1 year", 1),
                (long.MaxValue, "years ago", "in years", 31536000)
            };

            foreach (var (max, singular, plural, divisor) in thresholds)
            {
                if (seconds <= max)
                {
                    if (singular == "just now")
                        return "just now";

                    if (divisor == 1)
                        return isFuture ? plural : singular;

                    long n = (long)Math.Round((double)seconds / divisor, MidpointRounding.AwayFromZero);

                    if (n == 1)
                    {
                        string unit = (isFuture ? plural : singular).Replace("ago", "").Replace("in ", "").Trim();
                        return isFuture ? $"in 1 {unit}" : $"1 {unit} ago";
                    }
                    else
                    {
                        string unit = (isFuture ? plural : singular).Replace("ago", "").Replace("in ", "").Trim();
                        return isFuture ? $"in {n} {unit}" : $"{n} {unit} ago";
                    }
                }
            }

            return ""; // Unreachable
        }

        // ============================================================================
        // duration: Formats seconds as a human-readable duration string
        // ============================================================================

        public class DurationOptions
        {
            public bool Compact { get; set; } = false;
            public int MaxUnits { get; set; } = 2;
        }

        public static string Duration(long seconds, DurationOptions? options = null)
        {
            if (seconds < 0)
                throw new ArgumentException("Seconds must not be negative");

            options ??= new DurationOptions();

            if (seconds == 0)
                return options.Compact ? "0s" : "0 seconds";

            var units = new (long size, string singular, string plural, string compact)[]
            {
                (31536000, "year", "years", "y"),
                (2592000, "month", "months", "mo"),
                (86400, "day", "days", "d"),
                (3600, "hour", "hours", "h"),
                (60, "minute", "minutes", "m"),
                (1, "second", "seconds", "s")
            };

            var parts = new List<string>();
            long remaining = seconds;

            for (int i = 0; i < units.Length; i++)
            {
                var (size, singular, plural, compact) = units[i];

                if (remaining >= size)
                {
                    long count;

                    // If this is the last unit slot, round instead of floor
                    if (parts.Count + 1 >= options.MaxUnits)
                    {
                        count = (long)Math.Round((double)remaining / size, MidpointRounding.AwayFromZero);
                        remaining = 0;
                    }
                    else
                    {
                        count = remaining / size;
                        remaining %= size;
                    }

                    if (options.Compact)
                    {
                        parts.Add($"{count}{compact}");
                    }
                    else
                    {
                        string label = count == 1 ? singular : plural;
                        parts.Add($"{count} {label}");
                    }

                    if (parts.Count >= options.MaxUnits)
                        break;
                }
            }

            return string.Join(options.Compact ? " " : ", ", parts);
        }

        // ============================================================================
        // parse-duration: Parses a duration string into seconds
        // ============================================================================

        public static long ParseDuration(string input)
        {
            input = input.Trim();

            if (string.IsNullOrEmpty(input))
                throw new ArgumentException("Empty string");

            if (input.StartsWith("-"))
                throw new ArgumentException("Negative duration");

            // Try colon notation first
            var colonMatch = Regex.Match(input, @"^(\d+):(\d{1,2})(?::(\d{1,2}))?$");
            if (colonMatch.Success)
            {
                int hours = int.Parse(colonMatch.Groups[1].Value);
                int minutes = int.Parse(colonMatch.Groups[2].Value);
                int seconds = colonMatch.Groups[3].Success ? int.Parse(colonMatch.Groups[3].Value) : 0;
                return hours * 3600L + minutes * 60L + seconds;
            }

            // Unit aliases map
            var unitMap = new Dictionary<string, long>
            {
                // years
                {"y", 31536000}, {"yr", 31536000}, {"yrs", 31536000}, {"year", 31536000}, {"years", 31536000},
                // months
                {"mo", 2592000}, {"month", 2592000}, {"months", 2592000},
                // weeks
                {"w", 604800}, {"wk", 604800}, {"wks", 604800}, {"week", 604800}, {"weeks", 604800},
                // days
                {"d", 86400}, {"day", 86400}, {"days", 86400},
                // hours
                {"h", 3600}, {"hr", 3600}, {"hrs", 3600}, {"hour", 3600}, {"hours", 3600},
                // minutes
                {"m", 60}, {"min", 60}, {"mins", 60}, {"minute", 60}, {"minutes", 60},
                // seconds
                {"s", 1}, {"sec", 1}, {"secs", 1}, {"second", 1}, {"seconds", 1}
            };

            // Normalize: lowercase, remove commas, strip "and", collapse whitespace
            string normalized = input.ToLower()
                .Replace(",", " ")
                .Replace(" and ", " ")
                .Trim();
            normalized = Regex.Replace(normalized, @"\s+", " ");

            // Match all number+unit pairs
            var matches = Regex.Matches(normalized, @"(\d+(?:\.\d+)?)\s*([a-z]+)");

            if (matches.Count == 0)
                throw new ArgumentException("No recognized number+unit pairs");

            double totalSeconds = 0;

            foreach (Match match in matches)
            {
                double value = double.Parse(match.Groups[1].Value, CultureInfo.InvariantCulture);
                string unit = match.Groups[2].Value;

                if (!unitMap.ContainsKey(unit))
                    throw new ArgumentException($"Unrecognized unit name");

                totalSeconds += value * unitMap[unit];
            }

            return (long)Math.Round(totalSeconds, MidpointRounding.AwayFromZero);
        }

        // ============================================================================
        // human-date: Contextual date string based on proximity
        // ============================================================================

        public static string HumanDate(long timestamp, long reference)
        {
            var tsDate = DateTimeOffset.FromUnixTimeSeconds(timestamp).UtcDateTime;
            var refDate = DateTimeOffset.FromUnixTimeSeconds(reference).UtcDateTime;

            // Truncate to midnight UTC
            var tsMidnight = new DateTime(tsDate.Year, tsDate.Month, tsDate.Day, 0, 0, 0, DateTimeKind.Utc);
            var refMidnight = new DateTime(refDate.Year, refDate.Month, refDate.Day, 0, 0, 0, DateTimeKind.Utc);

            int dayDiff = (int)Math.Round((tsMidnight - refMidnight).TotalDays);

            string[] dayNames = { "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday" };
            string[] monthNames = { "", "January", "February", "March", "April", "May", "June", "July",
                                   "August", "September", "October", "November", "December" };

            if (dayDiff == 0)
                return "Today";
            if (dayDiff == -1)
                return "Yesterday";
            if (dayDiff == 1)
                return "Tomorrow";
            if (dayDiff >= -6 && dayDiff <= -2)
                return $"Last {dayNames[(int)tsDate.DayOfWeek]}";
            if (dayDiff >= 2 && dayDiff <= 6)
                return $"This {dayNames[(int)tsDate.DayOfWeek]}";

            // Same year
            if (tsDate.Year == refDate.Year)
                return $"{monthNames[tsDate.Month]} {tsDate.Day}";

            // Different year
            return $"{monthNames[tsDate.Month]} {tsDate.Day}, {tsDate.Year}";
        }

        // ============================================================================
        // date-range: Smart date range formatting
        // ============================================================================

        public static string DateRange(long start, long end)
        {
            // Auto-swap if needed
            if (start > end)
                (start, end) = (end, start);

            var startDate = DateTimeOffset.FromUnixTimeSeconds(start).UtcDateTime;
            var endDate = DateTimeOffset.FromUnixTimeSeconds(end).UtcDateTime;

            string[] monthNames = { "", "January", "February", "March", "April", "May", "June", "July",
                                   "August", "September", "October", "November", "December" };

            // Same day
            if (startDate.Year == endDate.Year && startDate.Month == endDate.Month && startDate.Day == endDate.Day)
                return $"{monthNames[startDate.Month]} {startDate.Day}, {startDate.Year}";

            // Same month & year
            if (startDate.Year == endDate.Year && startDate.Month == endDate.Month)
                return $"{monthNames[startDate.Month]} {startDate.Day}\u2013{endDate.Day}, {startDate.Year}";

            // Same year, different month
            if (startDate.Year == endDate.Year)
                return $"{monthNames[startDate.Month]} {startDate.Day} \u2013 {monthNames[endDate.Month]} {endDate.Day}, {startDate.Year}";

            // Different years
            return $"{monthNames[startDate.Month]} {startDate.Day}, {startDate.Year} \u2013 {monthNames[endDate.Month]} {endDate.Day}, {endDate.Year}";
        }
    }
}
